src/main/java/org/cru/redegg/recording/log4j2/RedEggLog4j2Appender.java

```diff
@@ -0,0 +1,113 @@
+package org.cru.redegg.recording.log4j2;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.core.Filter;
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.appender.AbstractAppender;
+import org.apache.logging.log4j.core.filter.ThresholdFilter;
+import org.cru.redegg.recording.api.ErrorRecorder;
+import org.cru.redegg.recording.api.RecorderFactory;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.LogRecord;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * @author Matt Drees
+ */
+public class RedEggLog4j2Appender extends AbstractAppender
+{
+
+    private final RecorderFactory factory;
+    private final Set<String> ignoredLoggerNames;
+
+    public RedEggLog4j2Appender(RecorderFactory factory, Set<String> ignoredLoggerNames) {
+        super("red-egg-appender", null, null, false);
+        this.factory = checkNotNull(factory);
+        this.ignoredLoggerNames = checkNotNull(ignoredLoggerNames);
+        addFilter(ThresholdFilter.createFilter(Level.DEBUG, Filter.Result.ACCEPT, Filter.Result.DENY));
+    }
+
+    @Override
+    public void append(LogEvent event)
+    {
+        if (ignoredLoggerNames.contains(event.getLoggerName()))
+            return;
+
+        ErrorRecorder recorder = factory.getRecorder();
+
+        if (event.getLevel().intLevel() >= Level.WARN.intLevel()) {
+
+            Map<String, String> properties = event.getContextData().toMap();
+            for (Map.Entry<String, String> entry : properties.entrySet())
+            {
+                recorder.recordContext(entry.getKey(), entry.getValue());
+            }
+
+            recorder.recordContext("log4j NDC", event.getContextStack().asList());
+        }
+
+        recorder
+            .recordLogRecord(toLogRecord(event))
+            .sendReportIfNecessary();
+
+    }
+
+
+    private LogRecord toLogRecord(LogEvent event) {
+
+        LogRecord record = new LogRecord(toJulLevel(event.getLevel()), String.valueOf(event.getMessage()));
+
+        if (event.getThrown() != null)
+        {
+            record.setThrown(event.getThrown());
+        }
+        else if (event.getMessage().getThrowable() != null) { //TODO: check if this makes sense to do
+            record.setThrown(event.getMessage().getThrowable());
+        }
+
+        record.setMillis(event.getTimeMillis());
+        record.setLoggerName(event.getLoggerName());
+        record.setSourceClassName(event.getSource().getClassName());
```

Comment:
this throws an NPE when testing this locally in The Key. the log events it's processing don't have `event.getSource()`

Reply:
so far I've only checked regular log messages, not log messages with exceptions.

Reply:
```
AsyncLogger error handling event seq=6633, value='Logger=me.thekey.cas.ticket.registry.SelfCleaningTicketRegistryCleaner Level=INFO Message=Finished ticket cleanup.':
org.apache.logging.log4j.core.appender.AppenderLoggingException: An exception occurred processing Appender red-egg-appender
	at org.apache.logging.log4j.core.appender.DefaultErrorHandler.error(DefaultErrorHandler.java:75)
	at org.apache.logging.log4j.core.config.AppenderControl.handleAppenderError(AppenderControl.java:165)
	at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:158)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:129)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:120)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:84)
	at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:448)
	at org.apache.logging.log4j.core.async.AsyncLoggerConfig.asyncCallAppenders(AsyncLoggerConfig.java:129)
	at org.apache.logging.log4j.core.async.AsyncLoggerConfigDisruptor$Log4jEventWrapperHandler.onEvent(AsyncLoggerConfigDisruptor.java:111)
	at org.apache.logging.log4j.core.async.AsyncLoggerConfigDisruptor$Log4jEventWrapperHandler.onEvent(AsyncLoggerConfigDisruptor.java:97)
	at com.lmax.disruptor.BatchEventProcessor.processEvents(BatchEventProcessor.java:168)
	at com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:125)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
	at org.cru.redegg.recording.log4j2.RedEggLog4j2Appender.toLogRecord(RedEggLog4j2Appender.java:73)
	at org.cru.redegg.recording.log4j2.RedEggLog4j2Appender.append(RedEggLog4j2Appender.java:53)
	at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:156)
	... 10 more
```

Reply:
thanks, I'll add a null-check